src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,337 @@
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Guard.Assert(toBlockHeight >= fromBlockHeight);
+
+            var headersInCache = new List<ProvenBlockHeader>();
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int index = fromBlockHeight;
+
+                ProvenBlockHeader header = null;
+
+                var headersNotInCache = new List<int>();
+
+                do
+                {
+                    this.Cache.TryGetValue(index, out header);
+
+                    if (header != null)
+                    {
+                        headersInCache.Add(header);
+                    }
+                    else
+                    {
+                        headersNotInCache.Add(index);
+                    }
+
+                    index++;
+
+                } while (index <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+
+                if (headersNotInCache.Count > 0)
+                {
+                    var repositoryHeaders = new List<ProvenBlockHeader>();
+
+                    // Check the full range first
+                    if ((headersNotInCache.Count - 1) == toBlockHeight - fromBlockHeight)
+                    {
+                        List<ProvenBlockHeader> rangeHeaders = await this.provenBlockHeaderRepository.GetAsync(fromBlockHeight, toBlockHeight);
+
+                        index = fromBlockHeight;
+
+                        foreach(ProvenBlockHeader rangeHeader in rangeHeaders)
+                        {
+                            if (rangeHeader != null)
+                            {
+                                repositoryHeaders.Add(rangeHeader);
+                                this.Cache.AddOrUpdate(index, rangeHeader, rangeHeader.HeaderSize);
+                            }
+
+                            index++;
+                        }
+                    }
+                    else
+                    {
+                        // If not a full sequence then check individually.
+                        foreach (int headerNotInCache in headersNotInCache)
+                        {
+                            ProvenBlockHeader repositoryHeader = await this.provenBlockHeaderRepository.GetAsync(headerNotInCache).ConfigureAwait(false);
+
+                            if (repositoryHeader != null)
+                            {
+                                repositoryHeaders.Add(repositoryHeader);
+
+                                this.Cache.AddOrUpdate(headerNotInCache, repositoryHeader, repositoryHeader.HeaderSize);
+                            }
+                        }
+                    }
+
+                    if (repositoryHeaders.Count > 0)
+                        headersInCache.AddRange(repositoryHeaders);
```

Comment:
Headers should be consecutive. 

Please fix and add following test case: 

HeaderStore is asked for headers from 100 to 110. 
Headers 103 and 105 are in case, rest are not. 

Make sure method returns consecutive headers from 100 to 110